Automattic/wp-calypso
# Introduce a new connect a domain flow

#### Changes proposed in this Pull Request

* This PR will introduce a new flow for users to "map" a domain. We'll be using the new terminology "connect a domain" instead. See: pcYYhz-e4-p2#new-connect-a-domain-flow

Note that before wiring this in, we'll need to add in the ability to store which "mode" was used when doing the verification so that we can send the user back to the correct step in the flow if they need to re-try the steps to get their domain connected. I'll add this in a separate PR since it requires some back-end changes as well.

#### Testing instructions

Use a path like `http://calypso.localhost:3000/domains/mapping/a8ctest.com/setup/example.com`

Walk through the flow on both the "suggested" and "advanced" paths.

Try this with a domain that has the correct NS/A records set and one that doesn't have them correctly set. Make sure that the final screen shows the correct case for "Connected" or "Verifying" your domain.

Try appending a query string with a specific step slug like `?step=advanced_update`. This should take you directly to that step.

When the `showErrors=1` query string is added, this will trigger an immediate verification check and any errors will be shown on the "update" screens.

Example walk through the "suggested flow":


https://user-images.githubusercontent.com/1379730/128061091-14c2abba-fec4-4062-bff4-0d3c4a22744b.mov

Example walk through using the "advanced flow":

https://user-images.githubusercontent.com/1379730/128061511-30f21ea5-c015-424c-a624-3c4fc9656760.mov

Example showing direct link to the update step and showing errors:

https://user-images.githubusercontent.com/1379730/128064332-e1f4410c-79b2-4bec-869b-8e43009d8be1.mov


Start step for suggested flow:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 05 PM" src="https://user-images.githubusercontent.com/1379730/128061551-7857f9ec-f89d-47ab-b00e-d8b994946494.png">

Login step (same for both flows, except for subtitle):
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 14 PM" src="https://user-images.githubusercontent.com/1379730/128061657-ea8f70f6-b74c-4f9a-8166-6dc9d834d3b4.png">

Update step for suggested flow:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 22 PM" src="https://user-images.githubusercontent.com/1379730/128061752-01a23d19-b609-47c3-8ba7-dd853c53989c.png">

Update step showing one of the clipboard buttons in a "copied" state:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 33 PM" src="https://user-images.githubusercontent.com/1379730/128061782-c0d42181-d318-4da1-9097-6c25552c28d6.png">
 
"Connected" step; shown when the flow is completed and the DNS records have been set correctly:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 24 06 PM" src="https://user-images.githubusercontent.com/1379730/128062062-010cecda-b477-4f52-b45a-0421644435de.png">


"Verifying" step; shown when the flow is completed, but the DNS records aren't yet propagated:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 45 PM" src="https://user-images.githubusercontent.com/1379730/128061979-0a877e50-0e4b-4f9f-8def-ef5bada27fa0.png">

Start step for advanced flow:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 15 58 PM" src="https://user-images.githubusercontent.com/1379730/128062125-8fffd233-ef20-4708-b53f-3cabd41093e8.png">

Update step for advanced flow:
<img width="1101" alt="Screen Shot 2021-08-03 at 1 16 17 PM" src="https://user-images.githubusercontent.com/1379730/128062209-b53afec6-ddf6-4f86-b4df-af734e7ce55e.png">
